### PR TITLE
add preFormCheck at createAction and editAction 

### PR DESCRIPTION
**Pull request by nnmer**
add preFormCheck at createAction and editAction after the form was initiated and request mapped to form and object

I am targeting this branch, because {reason}.

```markdown
### Added
- Added `CRUDController::preFormCheck` to have easy way to do custom validation of the object and form data

```

## Subject

The situation is next. I submitted the form by "Submit" button. In controller before to do save I want to do some custom validation for the object and submitted data.

Yes, we have `preCreate` and `preEdit` where arguments are $request and $object.
First: why $request there? we can any time do $this->getRequest(), at some point this parameter is useless.
Second: inside this method $object is just instantiated, and basically most of the properties are `null`, and if i need to validate form data, i need to copy 
```php
$this->admin->setSubject($object);

/** @var $form \Symfony\Component\Form\Form */
$form = $this->admin->getForm();
$form->setData($object);
$form->handleRequest($request);
``` 
into `preCreate`, so then i can check the Object and Form and maybe add explicit Form Error.
Then after `if ($form->isSubmitted()) {` and  `$isFormValid = $form->isValid()`   i will get the response with the form and Error message in it. Or return the response.

So, this copy/past action hopefully can be avoided.
A proposition is to add `preCormCheck` so inside we can access mapped data into the $form and $object.

Can be argued that we have `$this->admin->preValidate($object)`. This method has no access to the Form, so cannot set any errors to the Form. 

if we modify 
```$this->admin->preValidate($object)```
into:

```php
$preResponse = $this->admin->preValidate($object, $form);
if ($preResponse !== null) {
    return $preResponse;
}
```
this also can solve a problem

### CODE REVIEW

File: Controller/CRUDController.php [hunk 1]
```diff
@@ -251,11 +251,16 @@ public function editAction($id = null)
 
         $this->admin->setSubject($object);
 
-        /** @var $form Form */
+        /** @var $form \Symfony\Component\Form\Form */
```
Comment (core23) [inline Controller/CRUDController.php]: No FQCN please

File: Controller/CRUDController.php [hunk 2]
```diff
@@ -1298,6 +1308,21 @@ protected function getCsrfToken($intention)
...
+     * It's called from createAction and editAction.
+     *
+     * @param Form $form
+     * @param mixed   $object
+     *
+     * @return Response|null
+     */
+    protected function preFormCheck(Form $form, $object)
+    {
+
```
Comment (core23) [inline Controller/CRUDController.php]: `return` is missing according to the PHPDoc

File: Resources/doc/reference/saving_hooks.rst [hunk 3]
```diff
@@ -119,6 +119,7 @@ redefine the following methods:
 
 - new object : ``preCreate($object)``
 - edited object : ``preEdit($object)``
+- new/edited object: ``preFormCheck($form, $object)
```
Comment (pulzarraider) [inline Resources/doc/reference/saving_hooks.rst]: Missing ending ``.

File: Controller/CRUDController.php [hunk 4]
```diff
@@ -1298,6 +1308,21 @@ protected function getCsrfToken($intention)
         return false;
     }
 
+
+    /**
+     * This method can be overloaded in your custom CRUD controller.
+     * It's called from createAction and editAction.
+     *
+     * @param Form $form
+     * @param mixed   $object
```
Comment (pulzarraider) [inline Controller/CRUDController.php]: Three spaces between `mixed` and `$object`

File: Controller/CRUDController.php (hunk 2 above)
Comment (nnmer) [inline Controller/CRUDController.php]: same as all other pre*** , they just a empty methods, i just follow

File: Controller/CRUDController.php (hunk 1 above)
Comment (nnmer) [inline Controller/CRUDController.php]: okay okay, actualy previously createAction had FQCN while editAction no

File: Resources/doc/reference/saving_hooks.rst (hunk 3 above)
Comment (nnmer) [inline Resources/doc/reference/saving_hooks.rst]: updated

File: Controller/CRUDController.php (hunk 4 above)
Comment (nnmer) [inline Controller/CRUDController.php]: updated

File: Controller/CRUDController.php (hunk 2 above)
Comment (core23) [inline Controller/CRUDController.php]: then we have a mismatch. 

Me return value is a `void` type

Comment (nnmer) [inline Controller/CRUDController.php]: see https://github.com/sonata-project/SonataAdminBundle/blob/3.x/Controller/CRUDController.php#L247
and https://github.com/sonata-project/SonataAdminBundle/blob/3.x/Controller/CRUDController.php#L1323

i formatted the method following this example

**Comment by OskarStark** [comment]
Please fix StyleCi

File: Controller/CRUDController.php [hunk 5]
```diff
@@ -1300,6 +1311,19 @@ protected function getCsrfToken($intention)
...
      * This method can be overloaded in your custom CRUD controller.
+     * It's called from createAction and editAction.
+     *
+     * @param Form  $form
+     * @param mixed $object
+     *
+     * @return Response|null
+     */
+    protected function preFormCheck(Form $form, $object)
+    {
```
Comment (core23) [inline Controller/CRUDController.php]: Referencing the PHPDoc, please return `null` by default

**Comment by core23** [comment]
This PR looks dead. Feel free to reopen if you will continue working on this.

**Pull request closed without merging**